File: backend/app/holdings/schemas.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.data_sources.domain import AssetType, MarketStatus
from app.overview.schemas import DataSourceSummary

HoldingStatus = Literal["open", "closed"]


class InstrumentResponse(BaseModel):
    thscode: str
    ticker: str
    name: str
    asset_type: AssetType
    exchange: Literal["SH", "SZ", "BJ"]
    industry: str | None = None


class InstrumentSearchResponse(BaseModel):
    items: list[InstrumentResponse]


class HoldingCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    thscode: str = Field(min_length=4, max_length=20)
    average_cost: Decimal = Field(ge=0, max_digits=20, decimal_places=4)
    quantity: int = Field(gt=0)
    opened_on: date
    note: str | None = Field(default=None, max_length=1000)

    @field_validator("thscode")
    @classmethod
    def normalize_thscode(cls, value: str) -> str:
        return value.upper()

    @field_validator("opened_on")
    @classmethod
    def opened_on_cannot_be_future(cls, value: date) -> date:
        if value > date.today():
            raise ValueError("建仓日期不得晚于今天")
        return value

    @field_validator("note")
    @classmethod
    def normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class HoldingUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    average_cost: Decimal | None = Field(
        default=None,
        ge=0,
        max_digits=20,
        decimal_places=4,
    )
    quantity: int | None = Field(default=None, ge=0)
    opened_on: date | None = None
    note: str | None = Field(default=None, max_length=1000)
    close_price: Decimal | None = Field(
        default=None,
        gt=0,
        max_digits=20,
        decimal_places=4,
    )
    closed_on: date | None = None
    closed_quantity: int | None = Field(default=None, gt=0)

    @field_validator("closed_on")
    @classmethod
    def closed_on_cannot_be_future(cls, value: date | None) -> date | None:
        if value is not None and value > date.today():
            raise ValueError("closed_on cannot be in the future")
        return value

    @field_validator("opened_on")
    @classmethod
    def opened_on_cannot_be_future(cls, value: date | None) -> date | None:
        if value is not None and value > date.today():
            raise ValueError("建仓日期不得晚于今天")
        return value

    @field_validator("note")
    @classmethod
    def normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @model_validator(mode="after")
    def require_submitted_field(self) -> "HoldingUpdate":
        if not self.model_fields_set:
            raise ValueError("请至少提交一个需要修改的字段")
        for field_name in (
            "average_cost",
            "quantity",
            "opened_on",
            "close_price",
            "closed_on",
            "closed_quantity",
        ):
            if field_name in self.model_fields_set and getattr(self, field_name) is None:
                raise ValueError(f"{field_name} 不能为 null")
        return self


class HoldingOrderPayload(BaseModel):
    holding_ids: list[uuid.UUID] = Field(default_factory=list)


class HoldingItem(BaseModel):
    id: uuid.UUID
    thscode: str
    ticker: str
    name: str
    asset_type: AssetType
    exchange: str
    average_cost: float
    quantity: int
    opened_on: date
    note: str | None
    sort_order: int
    status: HoldingStatus
    closed_quantity: int | None
    close_price: float | None
    closed_on: date | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    cost_amount: float
    close_amount: float | None = None
    realized_gain: float | None = None
    realized_gain_percent: float | None = None
    latest: float | None = None
    market_value: float | None = None
    floating_gain: float | None = None
    floating_gain_percent: float | None = None
    change_percent: float | None = None
    weight_percent: float | None = None
    quoted_at: datetime | None = None


class HoldingsListResponse(BaseModel):
    status: HoldingStatus
    items: list[HoldingItem]
    data_source: DataSourceSummary
    market_status: MarketStatus
    polling_enabled: bool
    refresh_seconds: int
    stale: bool = False


class HoldingSummaryResponse(BaseModel):
    total_cost: float
    priced_cost: float
    total_market_value: float | None
    floating_gain: float | None
    floating_gain_percent: float | None
    incomplete: bool
    holding_count: int
    realized_gain: float | None
    realized_gain_percent: float | None
    realized_incomplete: bool
    total_gain: float | None
    data_source: DataSourceSummary
    market_status: MarketStatus
    polling_enabled: bool
    refresh_seconds: int
    stale: bool = False


class MessageResponse(BaseModel):
    message: str

File: .pnpm-config/pnpm/store/v11/projects/d2bd198cdd43630583dbbe9675e8d12c/src/features/holdings/schemas.ts
import { z } from "zod";

const dataSourceSummarySchema = z
  .object({
    state: z.enum(["ready", "not_configured", "authentication_failed", "rate_limited", "unavailable"]),
    name: z.string().nullable(),
    message: z.string().nullable(),
  })
  .passthrough();

const marketStatusSchema = z.enum(["交易中", "午间休市", "已收盘", "休市", "未知"]);
const assetTypeSchema = z.enum(["a_share", "fund_etf"]);
const holdingStatusSchema = z.enum(["open", "closed"]);

export const portfolioSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    note: z.string().nullable(),
    is_default: z.boolean(),
    sort_order: z.number(),
    open_holding_count: z.number(),
    created_at: z.string(),
    updated_at: z.string(),
  })
  .passthrough();

export const portfolioListSchema = z
  .object({
    items: z.array(portfolioSchema),
  })
  .passthrough();

export const holdingSchema = z
  .object({
    id: z.string(),
    thscode: z.string(),
    ticker: z.string(),
    name: z.string(),
    asset_type: assetTypeSchema,
    exchange: z.string(),
    average_cost: z.number(),
    quantity: z.number(),
    opened_on: z.string(),
    note: z.string().nullable(),
    sort_order: z.number(),
    status: holdingStatusSchema,
    closed_quantity: z.number().nullable(),
    close_price: z.number().nullable(),
    closed_on: z.string().nullable(),
    closed_at: z.string().nullable(),
    created_at: z.string(),
    updated_at: z.string(),
    cost_amount: z.number(),
    close_amount: z.number().nullable(),
    realized_gain: z.number().nullable(),
    realized_gain_percent: z.number().nullable(),
    latest: z.number().nullable(),
    market_value: z.number().nullable(),
    floating_gain: z.number().nullable(),
    floating_gain_percent: z.number().nullable(),
    change_percent: z.number().nullable(),
    weight_percent: z.number().nullable(),
    quoted_at: z.string().nullable(),
  })
  .passthrough();

export const holdingsListSchema = z
  .object({
    status: holdingStatusSchema,
    items: z.array(holdingSchema),
    data_source: dataSourceSummarySchema,
    market_status: marketStatusSchema,
    polling_enabled: z.boolean(),
    refresh_seconds: z.number(),
    stale: z.boolean(),
  })
  .passthrough();

export const holdingSummarySchema = z
  .object({
    total_cost: z.number(),
    priced_cost: z.number(),
    total_market_value: z.number().nullable(),
    floating_gain: z.number().nullable(),
    floating_gain_percent: z.number().nullable(),
    incomplete: z.boolean(),
    holding_count: z.number(),
    realized_gain: z.number().nullable(),
    realized_gain_percent: z.number().nullable(),
    realized_incomplete: z.boolean(),
    total_gain: z.number().nullable(),
    data_source: dataSourceSummarySchema,
    market_status: marketStatusSchema,
    polling_enabled: z.boolean(),
    refresh_seconds: z.number(),
    stale: z.boolean(),
  })
  .passthrough();

export const portfolioSummaryListSchema = z
  .object({
    items: z.array(
      z
        .object({
          portfolio_id: z.string(),
          summary: holdingSummarySchema,
        })
        .passthrough(),
    ),
  })
  .passthrough();

export const instrumentSearchSchema = z
  .object({
    items: z.array(
      z
        .object({
          thscode: z.string(),
          ticker: z.string(),
          name: z.string(),
          asset_type: assetTypeSchema,
          exchange: z.enum(["SH", "SZ", "BJ"]),
          industry: z.string().nullable().optional(),
        })
        .passthrough(),
    ),
  })
  .passthrough();
